### PR TITLE
Fixed the FaceAlias processes

### PR DESCRIPTION
**Pull request by geralddejong**
Querying, augmenting, filtering

### CODE REVIEW

File: src/build/brick.rs [hunk 1]
```diff
@@ -318,13 +318,20 @@ impl Baked {
...
+        let baked = &baked_bricks
             .iter()
-            .filter(|(baked_alias, _)| alias.matches(baked_alias))
+            .filter(|(baked_alias, _)| search_base.matches(baked_alias))
             .min_by_key(|(brick_alias, _)| brick_alias.0.len())
-            .expect(&format!("no such brick: '{alias}'"))
-            .1
-            .clone()
+            .expect(&format!("no such brick: '{search_base}'")).1;
+        let mut thawed = baked.clone();
```
Comment (mitchelljustin) [inline src/build/brick.rs:327]: "Thawed" lol, nice

File: src/build/brick.rs [hunk 2]
```diff
@@ -318,13 +318,20 @@ impl Baked {
...
+        for face in &mut thawed.faces {
+            face.aliases.retain(|candidate| search_alias.matches(candidate));
+            if face.aliases.len() != 1 {
+                panic!("exactly one face should be retained");
```
Comment (mitchelljustin) [inline src/build/brick.rs:331]: I'm starting to feel we throw panic!s and unwraps() around too cavalierly. We should do a panic-to-Result refactor in the future

File: src/build/brick.rs [hunk 3]
```diff
@@ -177,7 +178,6 @@ impl Prototype {
                 _ => unreachable!("{:?}", pair.as_rule()),
             }
         }
-        // TODO: validate all the names used
```
Comment (mitchelljustin) [inline src/build/brick.rs:180]: Don't wanna do this anymore?

File: src/build/brick.rs [hunk 4]
```diff
@@ -291,7 +291,7 @@ impl Baked {
         }
     }
 
-    pub fn new_brick(alias: &FaceAlias) -> Baked {
+    pub fn new_brick(search_alias: &FaceAlias) -> Baked {
```
Comment (mitchelljustin) [inline src/build/brick.rs:294]: 👌🏻

File: src/build/brick.rs [hunk 5]
```diff
@@ -318,13 +318,20 @@ impl Baked {
...
+        let search_base = search_alias.with_base();
+        let baked = &baked_bricks
             .iter()
-            .filter(|(baked_alias, _)| alias.matches(baked_alias))
+            .filter(|(baked_alias, _)| search_base.matches(baked_alias))
             .min_by_key(|(brick_alias, _)| brick_alias.0.len())
-            .expect(&format!("no such brick: '{alias}'"))
-            .1
-            .clone()
+            .expect(&format!("no such brick: '{search_base}'")).1;
```
Comment (mitchelljustin) [inline src/build/brick.rs:326]:
```suggestion
            .map(|(_, brick)| *brick)
            .expect(&format!("no such brick: '{search_base}'"));
```
This makes the unpacking of the tuple clearer than postfixing a `.1`

File: src/build/brick.rs (hunk 2 above)
Comment (geralddejong) [reply src/build/brick.rs:331]: I think this one is just an assertion, but yes.

File: src/build/brick.rs (hunk 3 above)
Comment (geralddejong) [reply src/build/brick.rs:180]: What was the validation?

Comment (mitchelljustin) [reply src/build/brick.rs:180]: Maybe not important enough

File: src/build/brick.rs [hunk 6]
```diff
@@ -318,13 +318,21 @@ impl Baked {
...
-            .1
-            .clone()
+            .map(|(_, brick)| *brick)
+            .expect(&format!("no such brick: '{search_base}'"));
+        let mut thawed = baked.clone();
+        for face in &mut thawed.faces {
+            face.aliases.retain(|candidate| search_alias.matches(candidate));
+            if face.aliases.len() != 1 {
+                panic!("exactly one face should be retained");
+            }
```
Comment (mitchelljustin) [inline src/build/brick.rs:333]:
```suggestion
            assert_eq(face.alias.len(), 1, "exactly one face should be retained");
```
In that case

File: src/fabric/brick.rs [hunk 7]
```diff
@@ -61,14 +57,21 @@ impl Fabric {
...
                 });
-                // TODO: make sure there's only one face left
-                self.create_face(aliases, scale, spin, radial_intervals)
+                let single_alias:Vec<_> = aliases
+                    .into_iter()
+                    .filter(|alias| search_alias.matches(alias))
+                    .collect();
+                if single_alias.len() != 1 {
+                    panic!("filter must leave exactly one face alias")
+                }
```
Comment (mitchelljustin) [inline src/fabric/brick.rs:66]:
```suggestion
                assert_eq!(single_alias.len(), 1, "filter must leave exactly one face alias");
```

File: src/fabric/brick.rs [hunk 8]
```diff
@@ -61,14 +57,19 @@ impl Fabric {
                     let ideal = self.ideal(alpha_index, omega_index, Baked::TARGET_FACE_STRAIN);
                     self.create_interval(alpha_index, omega_index, Link::pull(ideal))
                 });
-                // TODO: make sure there's only one face left
-                self.create_face(aliases, scale, spin, radial_intervals)
+                let single_alias:Vec<_> = aliases
```
Comment (mitchelljustin) [inline src/fabric/brick.rs:60]:
```suggestion
                let single_alias: Vec<_> = aliases
```

File: src/fabric/brick.rs [hunk 9]
```diff
@@ -61,14 +57,19 @@ impl Fabric {
...
+                let single_alias: Vec<_> = aliases
+                    .into_iter()
+                    .filter(|alias| search_alias.matches(alias))
+                    .collect();
+                assert_eq!(single_alias.len(), 1, "filter must leave exactly one face alias");
+                self.create_face(single_alias, scale, spin, radial_intervals)
             })
             .collect();
         if let Some(id) = face_id {
+            let search_base = search_alias.with_base();
```
Comment (mitchelljustin) [inline src/fabric/brick.rs:69]:
```suggestion
            let search_alias = search_alias.with_base();
```
search_base is a more unclear name to me, so just shadow "search_alias"

**Review by mitchelljustin — APPROVED** [review]

**Pull request merged**